grapher/mapCharts/AnnotationGenerator.ts

```diff
@@ -0,0 +1,922 @@
+import { PointVector } from "../../clientUtils/PointVector.js"
+import pixelWidth from "string-pixel-width"
+import {
+    ChoroplethSeries,
+    InternalLabel,
+    RenderFeature,
+} from "./MapChartConstants.js"
+import { MapProjectionName } from "./MapProjections.js"
+import { geoPath } from "d3-geo"
+import polylabel from "polylabel"
+import { Position } from "geojson"
+import { Bounds } from "../../clientUtils/Bounds.js"
+
+enum ExternalPositions {
+    right = "right",
+    left = "left",
+    topRight = "topRight",
+    bottomRight = "bottomRight",
+    topLeft = "topLeft",
+    bottomLeft = "bottomLeft",
+    bottom = "bottom",
+    top = "top",
+}
+
+interface ExternalArea {
+    id: string
+    region: Position[]
+    area: number
+    pole: Position
+    regionPoints: Position[]
+}
+
+interface InternalPosition {
+    position: number[]
+    points: Position[]
+    id: string
+}
+
+interface CoastPositions {
+    positions: { [position: string]: Position }
+    id: string
+}
+interface CalculatedPosition {
+    id: string
+    boundaryPosition: Position
+    labelPosition: string
+    textWidth: number
+    noHope: boolean
+    finalPosition?: Position
+    markerStart?: Position
+    markerEnd?: Position
+    anchor?: boolean
+}
+interface CombinedRegions {
+    id: string
+    points: Position[]
+}
+
+interface PositionsCacheForProjection {
+    coastPositions: CoastPositions[]
+    calculatedPositions: CalculatedPosition[]
+    combinedRegions: CombinedRegions[]
+    polylabelCache: InternalPosition[]
+    allPoints: Record<string, number>
+}
+
+let globalScale: number
+
+export function generateAnnotations(
+    featureData: RenderFeature[],
+    featuresWithNoData: RenderFeature[],
+    choroplethData: Map<string, ChoroplethSeries>,
+    viewportScale: number,
+    offset: number[],
+    bounds: Bounds,
+    projection: MapProjectionName,
+    annotationsCache: Map<string, any>
+): InternalLabel[] {
```

Comment:
The return type name is no longer really correct, right? Maybe a good name would be ValueLabel?